README.md
# Thea SDK [![npm version](https://badge.fury.io/js/@mcovilo%2Fthea-sdk.svg)](https://badge.fury.io/js/@mcovilo%2Fthea-sdk)

This is official JS SDK for Thea Protocol

## SDK Initialization

Depending on the platform you're using, there are different ways to initialize the SDK. Below are few samples

```js
// Client side
const theaSDK = await TheaSDK.init({
	network: TheaNetwork.MUMBAI,
	web3Provider: new ethers.providers.Web3Provider(window.ethereum)
});

// Server side with private key and provider
const theaSDK = await TheaSDK.init({
	network: TheaNetwork.MUMBAI,
	privateKey: "123...",
	provider: new ethers.providers.AlchemyProvider(80001, "apiKey")
});

// Server side with signer
const theaSDK = await TheaSDK.init({
	network: TheaNetwork.MUMBAI,
	signer: new Wallet(privateKey, provider)
});
```
You can also pass in a current NBT address as an optional parameter to the SDK initialization. If you don't pass in a current NBT address, the SDK will fetch current base token addres from BaseTokenManager SC. Here is an example of passing in a current NBT address as a parameter:

```js
// Server side init with signer and current NBT address
const theaSDK = await TheaSDK.init({
	network: TheaNetwork.MUMBAI,
	signer: new Wallet(privateKey, provider),
    currentNBTokenAddress: "0x123..."
});
```

- **Note!** You can set new `CurrentNBT` token address anytime by calling:

```js
theaSDK.setCurrentNBTContractAddress("0x123...");
```

## Convert module

- Convert NFTs - Converts `tokenAmount` of Carbon Credit ERC1155 with specific `tokenId` into corresponding amounts of current base token, SDG, Rating and Vintage ERC20's. BaseTokenManager contract gets ownership(locks) ERC1155's and mint's new ERC20 tokens to users address.

```js
// call to unwrap method of Registry contract
const result = await theaSDK.convert.convertNFT(tokenId, tokenAmount);

// Sample output
{
    "id": "1", // token ID
    "amount": '10', // Converted Amount of ERC1155 Tokens
    "to": '0xE100c4ffFF7c00253BA4A2a695F5ac909d756D76',
    "from": '0xbd44572e53343A0f003b719cf438C6338bD29d9C',
    ...
}
```

## Recover module

- Recover NFTs - Recovers `tokenAmount` of Carbon Credit ERC1155 with specific `tokenId` from corresponding amounts of current base token, SDG, Rating and Vintage ERC20's. BaseTokenManager contract sends to user ERC1155's and burns's ERC20 tokens in appropriate amonts.

```js
// call to unwrap method of Registry contract
const result = await theaSDK.recover.recoverNFT(tokenId, tokenAmount);

// Sample output
{
    "id": "1", // token ID
    "amount": '10', // Recovered Amount of ERC1155 Tokens
    "to": '0xE100c4ffFF7c00253BA4A2a695F5ac909d756D76',
    "from": '0xbd44572e53343A0f003b719cf438C6338bD29d9C',
    ...
}
```

- Query recover fungible - Returns list of token amounts needed to recover amount of specified NFT (by tokenID).

```js
// call to unwrap method of Registry contract
const result = await theaSDK.recover.queryRecoverFungibles(tokenId, tokenAmount);

// Sample output
{
    "cbt": "100000000", // current base token amount
    "sdg": '100000000', // sdg tokent amount
    "vintage": '100000000', // vintage token amount
    "rating": '100000000' // rating token amount
}
```

## Roll Tokens module

- Roll base tokens - Rolls `tokenAmount` of Base Token based on their `vintage`. This means that old Base Tokens (defined by sent `vintage`) and vintage tokens are burned and the same amount(`tokenAmount`) of new Base Tokens (defined by `vintage + 1`) is beeing minted. These Base Tokens addresses are defined in BaseTokenManager contract under `baseTokens` mapping with `vintage` as key.

```js
// call to unwrap method of Registry contract
const result = await theaSDK.rollBaseTokens.rollTokens(vintage, tokenAmount);

// Sample output
{
    "user": "0xbd44572e53343A0f003b719cf438C6338bD29d9C", // Address of user who rolled tokens
    "vintage": "2022", // Vintage of old Base Token
    "amount": '10', // Amount of base tokens rolled
    "to": '0xE100c4ffFF7c00253BA4A2a695F5ac909d756D76',
    "from": '0xbd44572e53343A0f003b719cf438C6338bD29d9C',
    ...
}
```

## Unwrap module

- Unwrap token - Stores a request to untokenize the VCC token of type `id`, locks the tokens and emits event. Backend listens to event and process the request. Tokens are not burnt until backend calls `updateUnwrapRequest` function after processing and validating the transfer to offchain `offchainAccount` was succesful.

```js
// call to unwrap method of Registry contract
const result = await theaSDK.unwrap.unwrapToken(tokenId, amount, offchainAccount)

// Sample output
{
    "requestId": "2", // Request ID that we can use to fetch token state
    "to": "0xe135783649BfA7c9c4c6F8E528C7f56166efC8a6", // The rest of the output is ethers Transaction receipt
    "from": "0xbd44572e53343A0f003b719cf438C6338bD29d9C",
    ...
}
```

- Get unwrap token state - Returns the state of the unwrap token request.

```js
// call to requests mapping of Registry contract
const state = await theaSDK.unwrap.getUnwrapTokenState(tokenId)

// Sample output
{
    "status": 0,
    "maker": "0x123...",
    "tokenId": "1",
    "amount": "1000"
}
```

## Tokenization module

- Request tokenization

```js
const tokenizationState = await theaSDK.tokenization.requestTokenization(
    TokenizationSource.VERRA,
    accountId,
    batchId,
    {
      email: "john@test.com",
      ethAddr: "0x123...",
      fullName: "John",
    }
  )
// Sample output
{
    "result": {
        "uuid": "0000018631954fb292bd329c70fe6751", // tokenization ID
        "createdAt": "2023-02-08T15:11:54.291Z",
        "updatedAt": "2023-02-08T15:11:54.291Z",
        "email": "john@test.com",
        "fullName": "John",
        "ethAddr": "0x123...",
        "source": "verra",
        "subaccountId": "43163",
        "batchId": "11158-2891187991...",
        "status": "IN_QUEUE",
        ...
    },
    "error": null,
    "errorMessage": null
}
```

- Get tokenization state - Query tokenization state by tokenization ID

```js
const tokenizationState = await theaSDK.tokenization.getTokenizationState(tokenizationId);
```

- Tokenize - Used to claim tokens after tokenization request is performed and admin approves it

```js
const tokenizationState = await theaSDK.tokenization.tokenize(tokenizationId);
```

## Offset module

- Offset NFT - offsets specified amount of ERC1155 token

```js
const transactionReceipt = await theaSDK.offset.offsetNFT(1, "2000");
```

- Offset fungible - offsets specified amount of ERC20 token for specified vintage

```js
// Returns requestId in transaction receipt
const transactionReceipt = await theaSDK.offset.offsetFungible(2017, "2000");
```

## Fungible trading

- Query token price - Used to fetch best price by calling Uniswap's V3 Quoter contract

```js
// Token in is "SDG" | "Vintage" | "Rating" | "CurrentNBT" | "Stable"

// From ERC20 to Stable coin
const priceInWEI = await theaSDK.fungibleTrading.queryTokenPrice({
	tokenIn: "SDG",
	amountIn: "1000000000000000000"
});

// From Stable coin to ERC20
const priceInWEI = await theaSDK.fungibleTrading.queryTokenPrice({
	tokenIn: "Stable",
	tokenOut: "SDG",
	amountIn: "1000000000000000000"
});
```

- Swap tokens - Used to swap tokens by calling Uniswap's V3 SwapRouter contract

```js
// From ERC20 to Stable coin
const transactionReceipt = await theaSDK.fungibleTrading.swapTokens({
	tokenIn: "SDG",
	amountIn: "1000000000000000000"
});

// From Stable coin to ERC20
const transactionReceipt = await theaSDK.fungibleTrading.swapTokens({
	tokenIn: "Stable",
	tokenOut: "SDG",
	amountIn: "1000000000000000000"
});
```

## NFT orderbook

- Query token list - Returns a list of ERC1155 tokens that are available, and they are grouped by projectIDs and vintages

```js
const priceInWEI = await theaSDK.nftTokenList.getTokenList();

// Sample output
// (Object which keys are projectIDs and values are arrays of tokens that belong to that project)
{
    "1784": // proojectId
    [       // list of tokens in that project
        {
            "vintage": 2018,
            "tokenID": "1",
            ...
        }
    ]
}
```

- Query Orders Info - Returns a list of orders for a given tokenID and owner

```js
const transactionReceipt = await theaSDK.nftOrderbook.queryOrdersInfo(tokenId, owner);

// Sample output
{
    "orders": [
        {
            "erc20Token": "0x5d29011d843b0b1760c43e10d66f302174bccd1a",
            "erc20TokenAmount": "10000000000000000000",
            "nftToken": "0xb47e3cd837ddf8e4c57f05d70ab865de6e193bbb",
            "nftTokenId": "69",
            "nftTokenAmount": "100",
            "nftType": "ERC1155",
            "sellOrBuyNft": "buy",
            "chainId": "1337",
            "order": {
                "direction": 1,
                "erc20Token": "0x5d29011d843b0b1760c43e10d66f302174bccd1a",
                "erc20TokenAmount": "10000000000000000000",
                "erc1155Token": "0xb47e3cd837ddf8e4c57f05d70ab865de6e193bbb",
                "erc1155TokenId": "69",
                "erc1155TokenAmount": "100",
                "erc1155TokenProperties": [],
                "expiry": "2524604400",
                "fees": [],
                "maker": "0x9342a65736a2e9c6a84a2adaba55ad1dc1f3a418",
                "nonce": "100131415900000000000000000000000000000096685863241593142117280893798097702934",
                "signature": {
                    "r": "0x39728a3bef397db69c6c6e1409ae6756c567a989894ad0787f9561113c9a80e9",
                    "s": "0x5f8a25be83efa2326e6405c68e8bdf5c0e83894dbef7e31de39d8c073302a1f6",
                    "v": 28,
                    "signatureType": 2
                },
                "taker": "0x0000000000000000000000000000000000000000"
            },
            "orderStatus": {
                "status": null, // null status indicates that order is not yet executed and is awailable for matching
                "transactionHash": null,
                "blockNumber": null
            },
            "metadata": {}
        },
        ...
    ]
}

```

- Query Price Listings - Returns sorted price listing for a given tokenID and side (buy or sell)

```js
const transactionReceipt = await theaSDK.nftOrderbook.queryPriceListing(tokenId, side);

// Sample output
[
	{
		"priceForOneNFT": 10,
		"nftTokenAmount": "10",
		"orderId": "100133271337000000000000000000000000000173830542377169720320941218856725572133",
		"orderToBeFilled": {
			"direction": 1,
			"erc20Token": "0xa6cbe96c05e92a01b52f519d50541409d85ed6d6",
			"erc20TokenAmount": "100000000000000000000",
			"erc1155Token": "0xf37221f42678ace417f2bc5c89489d1f0c77c133",
			"erc1155TokenId": "1",
			"erc1155TokenAmount": "10",
			"erc1155TokenProperties": [],
			"expiry": "2524604400",
			"fees": [],
			"maker": "0xca1edbea332fe36a4164bfc85bc58de12f07f941",
			"nonce": "100133271337000000000000000000000000000173830542377169720320941218856725572133",
			"signature": {
				"r": "0x004c3187cd41552901eeda3b6aa8bf1934db4cbff9e080d193b30714a62f9ead",
				"s": "0x293711b92b8dd052bdaebd00b84f2997e67b0b6ba8cd699ad513173f2856c867",
				"v": 28,
				"signatureType": 2
			},
			"taker": "0x0000000000000000000000000000000000000000"
		}
	},
    ...
]

```

- Query Order By Nonce - Returns order for a given nonce (orderId)

```js
const transactionReceipt = await theaSDK.nftOrderbook.queryOrderByNonce(nonce);

// Sample output
{
    "orders": [
        {
            "erc20Token": "0x5d29011d843b0b1760c43e10d66f302174bccd1a",
            "erc20TokenAmount": "10000000000000000000",
            "nftToken": "0xb47e3cd837ddf8e4c57f05d70ab865de6e193bbb",
            "nftTokenId": "69",
            "nftTokenAmount": "100",
            "nftType": "ERC1155",
            "sellOrBuyNft": "buy",
            "chainId": "1337",
            "order": {
                "direction": 1,
                "erc20Token": "0x5d29011d843b0b1760c43e10d66f302174bccd1a",
                "erc20TokenAmount": "10000000000000000000",
                "erc1155Token": "0xb47e3cd837ddf8e4c57f05d70ab865de6e193bbb",
                "erc1155TokenId": "69",
                "erc1155TokenAmount": "100",
                "erc1155TokenProperties": [],
                "expiry": "2524604400",
                "fees": [],
                "maker": "0x9342a65736a2e9c6a84a2adaba55ad1dc1f3a418",
                "nonce": "100131415900000000000000000000000000000096685863241593142117280893798097702934",
                "signature": {
                    "r": "0x39728a3bef397db69c6c6e1409ae6756c567a989894ad0787f9561113c9a80e9",
                    "s": "0x5f8a25be83efa2326e6405c68e8bdf5c0e83894dbef7e31de39d8c073302a1f6",
                    "v": 28,
                    "signatureType": 2
                },
                "taker": "0x0000000000000000000000000000000000000000"
            },
            "orderStatus": {
                "status": null, // null status indicates that order is not yet executed and is awailable for matching
                "transactionHash": null,
                "blockNumber": null
            },
            "metadata": {}
        }
    ]
}
```

## NFT trading

- Enter Limit Order - Creates and submits an off-chain limit order (for specific `tokenId`, `side`, `price` and `quantity`) to the orderbook

```js
const priceInWEI = await theaSDK.nftTrading.enterNFTLimit(
		tokenId, // token id of the NFT
		side, // buy or sell
		price, // price of the NFT in Stablecoins
		quantity, // quantity of the NFT
		chunks? // number of chunks to split the order into (default 1), each chunk will be made into a separate limit order
	);

// Sample output
{
    "erc20Token": "0xa6fa4fb5f76172d178d61b04b0ecd319c5d1c0aa",
    "erc20TokenAmount": "100000000000000",
    "nftToken": "0x2953399124f0cbb46d2cbacd8a89cf0599974963",
    "nftTokenId": "113604032257357238510303590891918450986076622282835488971632849699550347132938",
    "nftTokenAmount": "1",
    "nftType": "ERC1155",
    "sellOrBuyNft": "buy",
    "chainId": "80001",
    "order": {
        "direction": 1,
        "erc20Token": "0xa6fa4fb5f76172d178d61b04b0ecd319c5d1c0aa",
        "erc20TokenAmount": "100000000000000",
        "erc1155Token": "0x2953399124f0cbb46d2cbacd8a89cf0599974963",
        "erc1155TokenId": "113604032257357238510303590891918450986076622282835488971632849699550347132938",
        "erc1155TokenAmount": "1",
        "erc1155TokenProperties": [],
        "expiry": "2524604400",
        "fees": [],
        "maker": "0x1e45d9f1be883653ebd1c3311866ec8fab9ba74a",
        "nonce": "100133271337000000000000000000000000000272122256199440695915800875629318974968",
        "signature": {
            "r": "0x103be09b2a88d1c2916a0f85c177cc9f84af3373efd7442d770416a38003ba7b",
            "s": "0x22ac77d199c445c5cf093ef1b6aace677efef0256f6b0d88cbfa4ae0d02b17a3",
            "v": 27,
            "signatureType": 2
        },
        "taker": "0x0000000000000000000000000000000000000000"
    },
    "orderStatus": {
        "status": null, // null status indicates that order is not yet executed and is awailable for matching
        "transactionHash": null,
        "blockNumber": null
    },
    "metadata": {}
}
```

- Cancel Order - Cancels order for specific orderId(`nonce`) on 0x Exchange Conctract and returns transaction receipt

```js
const transactionReceipt = await theaSDK.nftTrading.cancelOrder(orderId);
```

- Update Order - Updates specific order with new `price` and `quantity`. Which includes cancels order for specific orderId(`nonce`) on 0x Exchange Conctract, then creating and submiting new off-chain limit order (for specific `tokenId` and `side` (defined in previously cancelled order), `price` and `quantity` (function input parameters)) to the orderbook

```js
const transactionReceipt = await theaSDK.nftTrading.updateOrder(orderId, price, quantity);

// Sample output
{
    "erc20Token": "0xa6fa4fb5f76172d178d61b04b0ecd319c5d1c0aa",

CHANGELOG.md
# Changelog

## [0.11.2](https://github.com/MVPWorkshop/thea-sdk/compare/v0.11.1...v0.11.2) (2023-02-23)


### Bug Fixes

* currentNBTokenAddress as a optional parameter for sdk initialization ([2f30e35](https://github.com/MVPWorkshop/thea-sdk/commit/2f30e350fb8d5dd1b5a6ca2121caaa563527e1d5))
* extra test removed ([e01397e](https://github.com/MVPWorkshop/thea-sdk/commit/e01397e2f3141d548ae30d75a64e761e688243de))
* get current NBT on thea sdk initialization ([cf79f9e](https://github.com/MVPWorkshop/thea-sdk/commit/cf79f9e11e0a8c56b13bbb5ec4120a788a52fae5))
* getERC20ContractAddress checck removed for currnet nbt ([056ac41](https://github.com/MVPWorkshop/thea-sdk/commit/056ac41ac66bf7abb4ac3015ab61ba096cedd17c))
* spender in roll tokens approve ([4a96d31](https://github.com/MVPWorkshop/thea-sdk/commit/4a96d314a73b4970f15ca2c50fad352f39ce03cf))

## [0.11.1](https://github.com/MVPWorkshop/thea-sdk/compare/v0.11.0...v0.11.1) (2023-02-23)


### Bug Fixes

* local addresses updated ([3c477d1](https://github.com/MVPWorkshop/thea-sdk/commit/3c477d14b062e28a9c6d3cc1f441f5632ace1a1a))

## [0.11.0](https://github.com/MVPWorkshop/thea-sdk/compare/v0.10.2...v0.11.0) (2023-02-23)


### Features

* offset fungible ([4a76353](https://github.com/MVPWorkshop/thea-sdk/commit/4a76353e5b726c8c4d2281b8be047e4a189e5a5d))

## [0.10.2](https://github.com/MVPWorkshop/thea-sdk/compare/v0.10.1...v0.10.2) (2023-02-22)


### Bug Fixes

* fix on handling errors from subgraph ([bed41c2](https://github.com/MVPWorkshop/thea-sdk/commit/bed41c228c4c8499eab12659a40aa4339bcaaa01))

## [0.10.1](https://github.com/MVPWorkshop/thea-sdk/compare/v0.10.0...v0.10.1) (2023-02-21)


### Bug Fixes

* documentation updated and test coverage raised ([2fb1e08](https://github.com/MVPWorkshop/thea-sdk/commit/2fb1e08112f37129c48b88f9c2e90096355845f4))
* get token list now uses subgraph ([a02927f](https://github.com/MVPWorkshop/thea-sdk/commit/a02927f8c063251a8f1439e0b94ba2f599ee6777))

## [0.10.0](https://github.com/MVPWorkshop/thea-sdk/compare/v0.9.5...v0.10.0) (2023-02-21)


### Features

* query recover fungibles ([ae9fe5b](https://github.com/MVPWorkshop/thea-sdk/commit/ae9fe5bbae81b5eb71fca90f509d94fab38c45e0))

## [0.9.5](https://github.com/MVPWorkshop/thea-sdk/compare/v0.9.4...v0.9.5) (2023-02-10)


### Bug Fixes

* return object include event parameters ([51dabce](https://github.com/MVPWorkshop/thea-sdk/commit/51dabce64368bc5ec5de0b929a2fec5da0961436))

## [0.9.4](https://github.com/MVPWorkshop/thea-sdk/compare/v0.9.3...v0.9.4) (2023-02-09)


### Bug Fixes

* address of stable coin on mumbai ([5526e1e](https://github.com/MVPWorkshop/thea-sdk/commit/5526e1ee9f380c1f434015e126372737e52df775))

## [0.9.3](https://github.com/MVPWorkshop/thea-sdk/compare/v0.9.2...v0.9.3) (2023-02-08)


### Bug Fixes

* thea api fix ([64b7c26](https://github.com/MVPWorkshop/thea-sdk/commit/64b7c26b3e4af2ce49fb3caf0520d6877573dcb3))

## [0.9.2](https://github.com/MVPWorkshop/thea-sdk/compare/v0.9.1...v0.9.2) (2023-02-08)


### Bug Fixes

* added missing module in sdk ([587265d](https://github.com/MVPWorkshop/thea-sdk/commit/587265da06689734118974b43999936da1e44226))

## [0.9.1](https://github.com/MVPWorkshop/thea-sdk/compare/v0.9.0...v0.9.1) (2023-02-08)


### Bug Fixes

* fix on setting current nbt contract address ([76cfb67](https://github.com/MVPWorkshop/thea-sdk/commit/76cfb67c341ee21252046220c1ac29048dd157af))
* token addreses updated ([5ef93e5](https://github.com/MVPWorkshop/thea-sdk/commit/5ef93e50aed9f9ac66d47dd5d09a1ffb89f36eaf))

## [0.9.0](https://github.com/MVPWorkshop/thea-sdk/compare/v0.8.1...v0.9.0) (2023-02-08)


### Features

* create limit orders in chunks ([f755fb0](https://github.com/MVPWorkshop/thea-sdk/commit/f755fb073aa25ee640e2107cc6ebecea6d820506))

## [0.8.1](https://github.com/MVPWorkshop/thea-sdk/compare/v0.8.0...v0.8.1) (2023-02-03)


### Bug Fixes

* init and provider network  mismatch ([86fdf0d](https://github.com/MVPWorkshop/thea-sdk/commit/86fdf0db5625490c840351307c05f9b6b3f2164c))

## [0.8.0](https://github.com/MVPWorkshop/thea-sdk/compare/v0.7.1...v0.8.0) (2023-02-03)


### Features

* estimate footprint calculator ([bf85a9b](https://github.com/MVPWorkshop/thea-sdk/commit/bf85a9b157b5fe14acb7b36c6921f084a386d613))

## [0.7.1](https://github.com/MVPWorkshop/thea-sdk/compare/v0.7.0...v0.7.1) (2023-02-02)


### Bug Fixes

* current nbt contract address hardcoded ([d7e58ba](https://github.com/MVPWorkshop/thea-sdk/commit/d7e58baac40d2a3413bb0063f422c1edadde202a))

## [0.7.0](https://github.com/MVPWorkshop/thea-sdk/compare/v0.6.0...v0.7.0) (2023-02-01)


### Features

* roll base tokens ([25425a9](https://github.com/MVPWorkshop/thea-sdk/commit/25425a908370daee5609d3d07c3369c4789f1190))

## [0.6.0](https://github.com/MVPWorkshop/thea-sdk/compare/v0.5.0...v0.6.0) (2023-01-31)


### Features

* amount check on swap and quote functions ([ec28946](https://github.com/MVPWorkshop/thea-sdk/commit/ec28946cd105c608cbfb38c9289f09c40e3fd0c7))


### Bug Fixes

* deadline check added ([8ac857b](https://github.com/MVPWorkshop/thea-sdk/commit/8ac857b65ed628aa219eb9d4ada050c5a7eda4c1))
* hardcoding pool for quote&swap ([6f7cef4](https://github.com/MVPWorkshop/thea-sdk/commit/6f7cef4595090421c7bc15cc143385ea741c96e0))

## [0.5.0](https://github.com/MVPWorkshop/thea-sdk/compare/v0.4.0...v0.5.0) (2023-01-30)


### Features

* nft trading module (without market order) ([e2957b6](https://github.com/MVPWorkshop/thea-sdk/commit/e2957b6d30fab9e707e6e8bc559f48b1e95cf365))
* quering token price from uniswap pools ([0b7f1dd](https://github.com/MVPWorkshop/thea-sdk/commit/0b7f1dd433a307ffd9eb680155dfe52d8a1a1bee))
* query order info ([e1a3579](https://github.com/MVPWorkshop/thea-sdk/commit/e1a357959d7bc4dc2f30672d3fd3c31f356ba71e))


### Bug Fixes

* chainID removed from env configs ([99bf60c](https://github.com/MVPWorkshop/thea-sdk/commit/99bf60c010dffcfd9e6735f0e63944404f03d2f4))

src/utils/consts.ts
import { BigNumber } from "@ethersproject/bignumber";
import { TheaNetwork } from "src/types";

export const RATE_VCC_TO_BT = 10;
export const STABLE_TOKEN_DECIMALS_MULTIPLIER = 10 ** 18;
export const ORDERBOOK_URL = "https://api.trader.xyz/orderbook";
export const DEFAULT_SLIPPAGE_TOLERANCE = 0.5;
export const INFINITE_EXPIRATION_TIMESTAMP_SEC = BigNumber.from(2524604400);
export const NULL_ADDRESS = "0x0000000000000000000000000000000000000000";
export const RESERVED_APP_ID_PREFIX = "1001";

// Some arbitrarily high number.
export const MAX_APPROVAL = BigNumber.from(2).pow(118);
// Number of digits in base 10 128bit nonce
// floor(log_10(2^128 - 1)) + 1
export const ONE_TWENTY_EIGHT_BIT_LENGTH = 39;
// Max nonce digit length in base 10
// floor(log_10(2^256 - 1)) + 1
export const TWO_FIFTY_SIX_BIT_LENGTH = 78;

export const ERC1155ORDER_STRUCT_NAME = "ERC1155Order";
export const ERC1155ORDER_STRUCT_ABI = [
	{ type: "uint8", name: "direction" },
	{ type: "address", name: "maker" },
	{ type: "address", name: "taker" },
	{ type: "uint256", name: "expiry" },
	{ type: "uint256", name: "nonce" },
	{ type: "address", name: "erc20Token" },
	{ type: "uint256", name: "erc20TokenAmount" },
	{ type: "Fee[]", name: "fees" },
	{ type: "address", name: "erc1155Token" },
	{ type: "uint256", name: "erc1155TokenId" },
	{ type: "Property[]", name: "erc1155TokenProperties" },
	{ type: "uint128", name: "erc1155TokenAmount" }
];

export const DEFAULT_APP_ID = "33271337";
export const FEE_ABI = [
	{ type: "address", name: "recipient" },
	{ type: "uint256", name: "amount" },
	{ type: "bytes", name: "feeData" }
];

export const PROPERTY_ABI = [
	{ type: "address", name: "propertyValidator" },
	{ type: "bytes", name: "propertyData" }
];
export enum Events {
	unwrap = "UnwrapRequested",
	convert = "Converted",
	recover = "Recovered",
	rollTokens = "Rolled",
	retireOffset = "RetireFungibleRequested"
}

export type EnvConfig = {
	networkName: string;
	registryContract: string;
	theaERC1155Contract: string;
	vintageTokenContract: string;
	sdgTokenContract: string;
	ratingTokenContract: string;
	currentNbtTokenContract: string;
	baseTokenManagerContract: string;
	baseTokenManagerDeployerContract: string;
	stableTokenContract: string;
	quoterContract: string;
	swapRouterContract: string;
	theaApiBaseUrl: string;
	exchangeProxyAddress: string;
	subGraphUrl: string;
};

export const consts: { [key in TheaNetwork]: EnvConfig } = {
	[TheaNetwork.GANACHE]: {
		networkName: "GANACHE",
		registryContract: "0x686AfD6e502A81D2e77f2e038A23C0dEf4949A20",
		theaERC1155Contract: "0xe135783649BfA7c9c4c6F8E528C7f56166efC8a6",
		vintageTokenContract: "0x686AfD6e502A81D2e77f2e038A23C0dEf4949A20",
		sdgTokenContract: "0x43D1F9096674B5722D359B6402381816d5B22F28",
		ratingTokenContract: "0x4261D524bc701dA4AC49339e5F8b299977045eA5",
		currentNbtTokenContract: "", // Call setCurrentNBTContractAddress to set address at init/thea.ts
		baseTokenManagerContract: "0x95C8f889701f20b624875a5188bEbDc9289b4F51",
		baseTokenManagerDeployerContract: "0x63491c5363329afb6f370E9D297025481E0277e6",
		stableTokenContract: "0x6B175474E89094C44Da98b954EedeAC495271d0F", //DAI
		quoterContract: "0xb27308f9F90D607463bb33eA1BeBb41C27CE5AB6",
		swapRouterContract: "0xE592427A0AEce92De3Edee1F18E0157C05861564",
		theaApiBaseUrl: "https://127.0.0.1:8078/cli",
		exchangeProxyAddress: "0xf91bb752490473b8342a3e964e855b9f9a2a668e",
		subGraphUrl: "http://localhost:8000/subgraphs/name/thea-protocol/thea-subgraph"
	},
	[TheaNetwork.MUMBAI]: {
		networkName: "MUMBAI",
		registryContract: "0x88449Dd0a1b75BC607A1E971b13930617D535EC1",
		theaERC1155Contract: "0xf37221F42678AcE417f2bc5c89489d1f0C77C133", // dummy erc1155
		vintageTokenContract: "0x724176146cd1fC5873480E4F2f53184B6734cbD9",
		sdgTokenContract: "0x4e5C34027912d6945B2C5CfE3A4E6CA16d50a482",
		ratingTokenContract: "0x007d5Af3e0AC91F3aEd13B866Ba05A0E4DBf8421",
		currentNbtTokenContract: "",
		baseTokenManagerContract: "0xE100c4ffFF7c00253BA4A2a695F5ac909d756D76",
		baseTokenManagerDeployerContract: "0x3ace09bba3b8507681146252d3dd33cd4e2d4f63",
		stableTokenContract: "0x014349F1C543038a76384cFC1A68f1881AFc6B0a", // Dummy USDC
		quoterContract: "0xb27308f9F90D607463bb33eA1BeBb41C27CE5AB6",
		swapRouterContract: "0xE592427A0AEce92De3Edee1F18E0157C05861564",
		theaApiBaseUrl: "https://127.0.0.1:8078/cli",
		exchangeProxyAddress: "0x4fb72262344034e034fce3d9c701fd9213a55260",
		subGraphUrl: "http://localhost:8000/subgraphs/name/thea-protocol/thea-subgraph"
	},
	[TheaNetwork.POLYGON]: {
		networkName: "POLYGON",
		registryContract: "0x88449Dd0a1b75BC607A1E971b13930617D535EC1",
		theaERC1155Contract: "0x22d5f9B75c524Fec1D6619787e582644CD4D7422", // sunflower
		vintageTokenContract: "0x3621027715647B69D706636a8878E85d725A2aed",
		sdgTokenContract: "0xB48C895039c9F81C87eb97Ed54B69a769b291f28",
		ratingTokenContract: "0xc95347BD5212148A09c34a7d890D061D73f50bb8",
		currentNbtTokenContract: "",
		baseTokenManagerContract: "0xE100c4ffFF7c00253BA4A2a695F5ac909d756D76",
		baseTokenManagerDeployerContract: "0x3ace09bba3b8507681146252d3dd33cd4e2d4f63",
		stableTokenContract: "0x2791Bca1f2de4661ED88A30C99A7a9449Aa84174",
		quoterContract: "0xb27308f9F90D607463bb33eA1BeBb41C27CE5AB6",
		swapRouterContract: "0xE592427A0AEce92De3Edee1F18E0157C05861564",
		theaApiBaseUrl: "https://127.0.0.1:8078/cli",
		exchangeProxyAddress: "0xdef1c0ded9bec7f1a1670819833240f027b25eff",
		subGraphUrl: "http://localhost:8000/subgraphs/name/thea-protocol/thea-subgraph"
	}
};

export type ISO_CODES =
	| "AFG"
	| "ALA"
	| "ALB"
	| "DZA"
	| "ASM"
	| "AND"
	| "AGO"
	| "AIA"
	| "ATA"
	| "ATG"
	| "ARG"
	| "ARM"
	| "ABW"
	| "AUS"
	| "AUT"
	| "AZE"
	| "BHS"
	| "BHR"
	| "BGD"
	| "BRB"
	| "BLR"
	| "BEL"
	| "BLZ"
	| "BEN"
	| "BMU"
	| "BTN"
	| "BOL"
	| "BES"
	| "BIH"
	| "BWA"
	| "BRA"
	| "VGB"
	| "BRN"
	| "BGR"
	| "BFA"
	| "BDI"
	| "KHM"
	| "CMR"
	| "CAN"
	| "CPV"
	| "CAF"
	| "TCD"
	| "CHL"
	| "CHN"
	| "CXR"
	| "COL"
	| "COM"
	| "COG"
	| "COK"
	| "CRI"
	| "CIV"
	| "HRV"
	| "CUB"
	| "CUW"
	| "CYP"
	| "CZE"
	| "COD"
	| "DNK"
	| "DJI"
	| "DMA"
	| "DOM"
	| "ECU"
	| "EGY"
	| "SLV"
	| "GNQ"
	| "ERI"
	| "EST"
	| "SWZ"
	| "ETH"
	| "FRO"
	| "FLK"
	| "FJI"
	| "FIN"
	| "FRA"
	| "GUF"
	| "PYF"
	| "GAB"
	| "GMB"
	| "GEO"
	| "DEU"
	| "GHA"
	| "GRC"
	| "GRL"
	| "GRD"
	| "GLP"
	| "GTM"
	| "GGY"
	| "GIN"
	| "GNB"
	| "GUY"
	| "HTI"
	| "HND"
	| "HKG"
	| "HUN"
	| "ISL"
	| "IND"
	| "IDN"
	| "IRN"
	| "IRQ"
	| "IRL"
	| "IMN"
	| "ISR"
	| "ITA"
	| "JAM"
	| "JPN"
	| "JEY"
	| "JOR"
	| "KAZ"
	| "KEN"
	| "KIR"
	| "KWT"
	| "KGZ"
	| "LAO"
	| "LVA"
	| "LBN"
	| "LSO"
	| "LBR"
	| "LBY"
	| "LIE"
	| "LTU"
	| "LUX"
	| "MAC"
	| "MDG"
	| "MWI"
	| "MYS"
	| "MDV"
	| "MLI"
	| "MLT"
	| "MHL"
	| "MTQ"
	| "MRT"
	| "MUS"
	| "MYT"
	| "MEX"
	| "FSM"
	| "MDA"
	| "MNG"
	| "MNE"
	| "MSR"
	| "MAR"
	| "MOZ"
	| "MMR"
	| "NAM"
	| "NRU"
	| "NPL"
	| "NLD"
	| "ANT"
	| "NCL"
	| "NZL"
	| "NIC"
	| "NER"
	| "NGA"
	| "NIU"
	| "PRK"
	| "MKD"
	| "NOR"
	| "OMN"
	| "PAK"
	| "PLW"
	| "PSE"
	| "PAN"
	| "PNG"
	| "PRY"
	| "PER"
	| "PHL"
	| "POL"
	| "PRT"
	| "PRI"
	| "QAT"
	| "REU"
	| "ROU"
	| "RUS"
	| "RWA"
	| "SHN"
	| "KNA"
	| "LCA"
	| "MAF"
	| "SPM"
	| "VCT"
	| "WSM"
	| "STP"
	| "SAU"
	| "SEN"
	| "SRB"
	| "SYC"
	| "SLE"
	| "SGP"
	| "SXM"
	| "SVK"
	| "SVN"
	| "SLB"
	| "SOM"
	| "ZAF"
	| "KOR"
	| "SSD"
	| "ESP"
	| "LKA"
	| "SDN"
	| "SUR"
	| "SJM"
	| "SWE"
	| "CHE"
	| "SYR"
	| "TWN"
	| "TJK"
	| "TZA"
	| "THA"
	| "TLS"
	| "TGO"
	| "TON"
	| "TTO"
	| "TUN"
	| "TUR"
	| "TKM"
	| "TCA"
	| "TUV"
	| "UGA"
	| "UKR"
	| "ARE"
	| "GBR"
	| "USA"
	| "VIR"
	| "URY"
	| "UZB"
	| "VUT"
	| "VEN"
	| "VNM"
	| "WLF"
	| "ESH"
	| "YEM"
	| "ZMB"
	| "ZWE";
